tests/platform_tests/fwutil/fwutil_common.py

```diff
@@ -0,0 +1,260 @@
+            return False
+    return True
+
+def call_fwutil(duthost, localhost, pdu_ctrl, fw, component=None, next_image=None, boot=None, basepath=None):
+    logger.info("Calling fwutil with component: {} | next_image: {} | boot: {} | basepath: {}".format(component, next_image, boot, basepath))
+    init_versions = show_firmware(duthost)
+    logger.info("Initial Versions: {}".format(init_versions))
+    chassis = init_versions["chassis"].keys()[0] # Only one chassis
+    paths = get_install_paths(duthost, fw, init_versions, chassis)
+    current = duthost.shell('sonic_installer list | grep Current | cut -f2 -d " "')['stdout']
+
+    generate_config(duthost, paths, init_versions)
+    upload_platform(duthost, paths, next_image)
+
+    command = "fwutil"
+    if basepath is not None:
+        command += " install"
+    else:
+        command += " update"
+
+    if component is None:
+        command += " all"
+    else:
+        if component not in paths:
+            pytest.skip("No available firmware to install on {}. Skipping".format(component))
+        command += " chassis component {} fw".format(component)
+
+    if basepath is not None:
+        # Install file is override if API implementation needs a different file for install / update
+        filepath = paths[component]["install"]["firmware"] if "install" in paths[component] else paths[component]["firmware"]
+        command += " {}".format(os.path.join(basepath, os.path.basename(filepath)))
+
+    if next_image is not None:
+        command += " --image={}".format("next" if next_image else "current")
+
+    if boot is not None:
+        command += " --boot={}".format(boot)
+
+    command += " -y"
+
+    logger.info("Running install command: {}".format(command))
+    task, res = duthost.command(command, module_ignore_errors=True, module_async=True)
+    boot_type = boot if boot else paths[component]["reboot"][0]
+
+    auto_reboot = False
+    for comp in paths.keys():
+        if "install" in paths[comp] and basepath is not None:
+            if paths[comp]["install"].get("auto_reboot", False): auto_reboot = True
+        else:
+            if paths[comp].get("auto_reboot", False): auto_reboot = True
+
+    timeout = max([v.get("timeout", TIMEOUT) for k, v in paths.items()])
+    pdu_delay = fw["chassis"][chassis].get("power_cycle_delay", 60)
+    complete_install(duthost, localhost, boot_type, res, pdu_ctrl, auto_reboot, current, next_image, timeout, pdu_delay)
+
+    final_versions = show_firmware(duthost)
+    assert validate_versions(init_versions, final_versions, paths, chassis, boot_type)
+
+    duthost.copy(src=os.path.join("firmware", "platform_components_backup.json"), 
+            dest=os.path.join(target, DEVICES_PATH, duthost.facts["platform"], "platform_components.json"))
+    logger.info("Restoring backup platform_components.json to {}".format(
+        os.path.join(DEVICES_PATH, duthost.facts["platform"])))
+
+    update_needed = copy(fw)
```

Comment:
what is this copy() doing?

Reply:
I perform `del` operations on update_needed in order to create a list of firmware which has been left at an older version by the test. 

This is destructive to the original data structure so I do a copy in order to get a ds which is not linked to the original `fw` object.